src/v/kafka/server/usage_manager.cc

```diff
@@ -0,0 +1,386 @@
+            last_window_delta = delta;
+        } else {
+            /// Close window and open a new one
+            buckets.at(*open_index).end = epoch_time_secs();
+            _current_window = (*open_index + 1) % buckets.size();
+            buckets.at(_current_window).reset(ss::lowres_system_clock::now());
+        }
+    }
+    _buckets = std::move(buckets);
+    return last_window_delta;
+}
+
+/// Operated on only by shard-0 this prevents execution of background work
+/// spawned by the configuration property update callbacks after stop has been
+/// called on the usage_manager
+static thread_local ss::gate bg_future_gate;
+
+usage_manager::usage_manager(ss::sharded<storage::api>& storage)
+  : _usage_enabled(config::shard_local_cfg().enable_usage.bind())
+  , _usage_num_windows(config::shard_local_cfg().usage_num_windows.bind())
+  , _usage_window_width_interval(
+      config::shard_local_cfg().usage_window_width_interval_sec.bind())
+  , _usage_disk_persistance_interval(
+      config::shard_local_cfg().usage_disk_persistance_interval_sec.bind())
+  , _storage(storage) {
+    _usage_enabled.watch([this] { (void)reset(); });
+    _usage_num_windows.watch([this] { (void)reset(); });
+    _usage_window_width_interval.watch([this] { (void)reset(); });
+    _usage_disk_persistance_interval.watch([this] { (void)reset(); });
+}
+
+ss::future<> usage_manager::reset() {
+    if (
+      !bg_future_gate.is_closed()
+      && (ss::this_shard_id() == usage_manager_main_shard)) {
+        auto h = bg_future_gate.hold();
+        /// Calling start() will send any new configuration updates to the
+        /// newly initialized accounting fiber
+        if (_accounting_fiber) {
+            auto fiber = std::move(_accounting_fiber);
+            _accounting_fiber = nullptr;
+            co_await fiber->stop().finally([h] {});
+        }
+        if (_usage_enabled()) {
+            co_await start();
+        }
+    }
+}
+
+ss::future<> usage_manager::start() {
+    if (ss::this_shard_id() != usage_manager_main_shard) {
+        co_return; /// Async work only occurs on shard-0
+    }
+    auto h = bg_future_gate.hold();
+    if (_accounting_fiber) {
+        co_return; /// Double start called, do-nothing
+    }
+    if (!_usage_enabled()) {
+        co_return; /// Feature is disabled in config
+    }
+    /// Reset all state
+    co_await container().invoke_on_all(
+      [](usage_manager& um) { (void)um.sample(); });
+
+    _accounting_fiber = std::make_unique<accounting_fiber>(
+      this->container(),
+      _storage,
+      _usage_num_windows(),
+      _usage_window_width_interval(),
+      _usage_disk_persistance_interval());
+    co_await _accounting_fiber->start().finally([h] {});
+}
+
+ss::future<> usage_manager::stop() {
+    if (!_accounting_fiber) {
+        return ss::now();
+    }
+    vassert(
+      ss::this_shard_id() == usage_manager_main_shard,
+      "_accounting fiber can only be non null on shard-0");
```

Comment:
this is confusing because it isn't symmetrical with the requirements of `start` which merely returns as a noop if called from the wrong thread.